Changes page Filters tab moved to React

benkeen opened the pull request:
This updates the Filter tab and filter on the Changes page
to use React. It also ports over Components.FilterView and
Components.FilterViewItem found in fauxton/components.js

benkeen commented:
Hey @robertkowalski, @garrensmith - my first foray into React. Woot!

I largely copied the structure etc. from the code that was already there, but I'm sure there's plenty to criticize. Mind taking a glance if you get any time today? Thanks!

robertkowalski commented:
looks promising!

 - we will need tests in order to merge it, one test could be that adding a filter adds it to the react wrapper component

 - I think you can delete the `Components.FilterView`

I'll add some other comments next to the code

robertkowalski commented:
When running the code I get:

```
Linting app/addons/documents/changes/changes-header.react.js ...ERROR
[L27:C8] W033: Missing semicolon.
```

robertkowalski commented:
stopping here for now - @garrensmith will talk to you later

benkeen commented on app/addons/fauxton/components.react.jsx:
```diff
@@ -150,12 +151,133 @@ function(FauxtonAPI, React, Stores, Actions) {
+      e.preventDefault();
+      e.stopPropagation();
+
+      var filter = this.refs.addItem.getDOMNode().value.trim();
```
Check this.

benkeen commented on app/addons/fauxton/components.react.jsx:
```diff
@@ -150,12 +151,133 @@ function(FauxtonAPI, React, Stores, Actions) {
+      var tooltip = '';
+      if (this.props.tooltip) {
+        tooltip = '<i ref="tooltip" className="icon icon-question-sign" data-toggle="tooltip" data-original-title=' +
+          this.props.tooltipText + '></i>';
```
move to helper

benkeen commented on app/addons/fauxton/components.react.jsx:
```diff
@@ -150,12 +151,133 @@ function(FauxtonAPI, React, Stores, Actions) {
+      var filters = _.map(this.state.filters, function (filter) {
+        return <FilterViewItem namespace={this.props.namespace} key={filter} label={filter} />;
+      }, this);
+
```
also move to helper

robertkowalski commented:
@benkeen this looks a lot better, we will need tests to merge this, examples:

- `ChangesHeader `: the element shows/removes the subcomponent conditionally if the state changes / actions are fired

- test what happens on `onToggle={this.toggleFilterSection}`

- are filters really removed after clicking on the remove button

benkeen commented:
Thanks, @robertkowalski! Just working on the tests now & ran into some trouble. I'll commit what I have, but I'll have questions tomorrow. Basically I'm unable to confirm that after entering text into the filter it gets added to the DOM. No matter how I try it, I can't find the elements in the updated DOM (possibly my assert() runs prior to the DOM being updated?). Committing now.

benkeen commented:
Alrighty! Tests added. I think this is good to go.

robertkowalski commented on app/addons/documents/changes/changes-header.react.jsx:
```diff
@@ -0,0 +1,236 @@
+      }
+
+      Actions.addFilterViewItem(filter);
+      this.refs.addItem.getDOMNode().value = '';
```
You are still directly accessing the DOM from an event handler?

Shouldn't we use an action connected to a store which then updates the element like this:

```
<input type="text" ref="addItem" placeholder={this.props.addItemPlaceholder} value="{this.state.addItemValue}" />
```

robertkowalski replied:
How does that relate to Flux? Should we always directly access the DOM instead of using the roundtrip using actions and stores?

garrensmith replied:
@robertkowalski is correct. This could be simplified to how we assigned a value from onChange

garrensmith replied:
Here are two examples of how we have worked with inputs using React:
[Index Editor](https://github.com/apache/couchdb-fauxton/blob/master/app/addons/documents/index-editor/components.react.jsx#L64)
[Cors](https://github.com/apache/couchdb-fauxton/blob/master/app/addons/cors/components.react.jsx#L76)

benkeen replied:
Thanks, guys. I gave this one some thought and I'm kind of on the fence. I'm more than happy making to the change, but it seems to add a lot of complexity just to make it more idiomatic.

If I understand it properly, I'd need to do the following:
- new onChange handler to DOM element
- handler which passing work to actions.js
- new event added in actiontypes.js
- state of input field value stored in the store
- a handler in the store to update the value as 

vs. what is it right now:
- one line within the react view setting the input value to ''.

I'm all for keeping the code as React-y as possible, but for me, this crosses the line. Ultimately, whole need to add in an onchange handler feels like a workaround to ensure state gets modified and that we can reset the value.

[Perfectly happy being overruled, here. Just my 2 cents :)]

garrensmith commented on app/addons/documents/changes/changes-header.react.jsx:
```diff
@@ -0,0 +1,236 @@
+    getDefaultProps: function () {
+      return {
+        addItemPlaceholder: 'Type a filter',
+        helpText: 'e.g. debug or <1.4.1> or any regex',
```
Could you add a better helpText. Why would anyone filter the changes feed with <1.4.1>?

garrensmith commented on app/addons/documents/changes/changes-header.react.jsx:
```diff
@@ -0,0 +1,236 @@
+    },
+
+    // props belong to the parent, so these aren't stored in the related store
+    getDefaultProps: function () {
```
I don't think we need these as props since they are never changed by the parent.

benkeen replied:
True. I just thought it would be good to design it in a generic way so that those could be easily changed down the road. But I guess I can remove them.

garrensmith commented on app/addons/documents/changes/stores.js:
```diff
@@ -0,0 +1,73 @@
+    },
+
+    reset: function () {
+      this.filters = [];
```
All internal variables for stores must be prefixed with an `_`. e.g `this._filters.

garrensmith commented on app/addons/documents/tests/changesHeader.componentsSpec.react.jsx:
```diff
@@ -0,0 +1,138 @@
+    });
+
+    it('should add filter markup', function () {
+      var $el = $(changesFilterEl.getDOMNode()),
```
This is cool. I like how you have broken the selecting of the DOM here. Makes the test a lot more readable.

garrensmith commented on app/addons/fauxton/components.react.jsx:
```diff
@@ -20,6 +20,7 @@ define([
 function(FauxtonAPI, React, Stores, Actions) {
   var navBarStore = Stores.navBarStore;
 
+
```
This is unrelated to the changes filter. I'm guessing this was by accident.

garrensmith commented on app/addons/fauxton/stores.js:
```diff
@@ -185,6 +185,7 @@ function(app, FauxtonAPI, ActionTypes) {
   Stores.navBarStore = new Stores.NavBarStore();
   Stores.navBarStore.dispatchToken = FauxtonAPI.dispatcher.register(Stores.navBarStore.dispatch);
 
+
```
This is unrelated to the changes filter. I'm guessing this was by accident.

benkeen commented:
Alrighty! *This* time it's totally good to go. ;) All suggested changes made.

garrensmith commented on app/addons/documents/changes/components.react.jsx:
```diff
@@ -0,0 +1,238 @@
+      e.preventDefault();
+      e.stopPropagation();
+
+      var filter = this.refs.addItem.getDOMNode().value.trim();
```
This needs to change to a [controlled controller](http://facebook.github.io/react/docs/forms.html)

garrensmith commented:
@benkeen making good progress. As I mentioned in the previous comment, the way the filter is added needs to be a controlled component.

benkeen commented:
Thanks @garrensmith. I think the input field was all set up properly - I just wasn't grabbing the current value from the store here. Updated.

benkeen commented:
Hey @garrensmith, I confess I was getting a bit confused about some things so I thought I'd just experiment a bit and organize things in what I regard as a logical manner. We can just revert it if I'm barking up the wrong tree.

Structure (pretty much the same):

```xml
<ChangesHeader />
	<ChangesHeaderTab />
	<ChangesFilter />
		<Filter /> (list of N components)
		<AddFilterForm />
			<AddFilterFormContent />
				<FilterTooltip />
```

Testing: these would be the key components, like before:
`<ChangesHeaderTab />` - test the hide/show of the tab
`<ChangesFilter />` - test the add/remove filters, tooltip content

Comments:

- It seemed weird that the components were all nice and discrete but shared the same store, so I split it into three: one for the header to track its hide/show state, one for the changes filter to track the list of visible filters, and one for the AddFilterForm to track the state of the text field. [Maybe a bit overengineered now?] I like the new arrangement because looking at a component you can see by checking its store what data is related to it. Before, I was losing track of what piece of state is associated with which component.
- I added a fauxton/mixins.js file to cut down on boilerplate code - namely the this.store.on('change', callback) assignment and destruction.
- I moved as much code relating to a particular component to the component itself, and stopped passing stuff down the chain. e.g. `<Filter />` now handles its own onRemove event and just fires an action. I'm not sure this is in keeping with the "dumb component" idea, but I find this way there's less dependencies between components and code is where I expect to find it.
- I'm still hazy on the need to separate `AddFilterForm` and the form content (`AddFilterFormContent`). They seem so functionally entwined, it feels a bit unnecessary - I'll have to get you to explain that one to me again.

benkeen commented:
Thanks for all your help, @garrensmith! All updated.
- actions + store tests added
- form moved back to a single component
- dropped third unnecessary store just to track the blank state of the AddFilterForm

garrensmith commented on app/addons/documents/tests/changes.componentsSpec.react.jsx:
```diff
@@ -0,0 +1,215 @@
+      assert.ok(Stores.changesHeaderStore.isTabVisible() === false);
+    });
+
+//    reset: function () {
```
You can probably just remove these

garrensmith commented on app/addons/documents/tests/changes.componentsSpec.react.jsx:
```diff
@@ -0,0 +1,215 @@
+      });
+    });
+
+    it('toggleTabVisibility() changes state in store', function() {
```
Can you move the store tests into their own stores test file. It doesn't make sense to have them in the components test file.

garrensmith commented on app/addons/documents/changes/components.react.jsx:
```diff
@@ -0,0 +1,244 @@
+
+    remove: function (e) {
+      e.preventDefault();
+      Actions.removeFilter(this.props.label);
```
The `<Filter />` component should be "dumb" and not call an action directly but call a callback passed to it from its controller component.

garrensmith commented on app/addons/documents/changes/components.react.jsx:
```diff
@@ -0,0 +1,244 @@
+      if (_.isEmpty(this.state.filter)) {
+        return;
+      }
+      Actions.addFilter(this.state.filter);
```
This should also be a dumb component that just calls a callback on a submit form and not an action directly.

garrensmith commented on app/addons/documents/changes/components.react.jsx:
```diff
@@ -0,0 +1,244 @@
+        <div className="tab-content">
+          <div className="tab-pane active" ref="filterTab">
+            <div className="changes-header js-filter">
+              <AddFilterForm tooltip={this.props.tooltip} filter={this.state.filter} onSubmit={this.submitForm} />
```
You don't have a `submitForm` method in this component.

garrensmith commented:
@benkeen this is looking better. Two components just need to be more self contained. I also picked up some bugs while testing. 

* Once I add a filter it doesn't clean the input field
* If I remove a fliter it doesn't regenerate the changes list without that filter
* Everytime I click on add I get this error `Warning: flattenChildren(...): Encountered two children with the same key, `.$hello`. Child keys must be unique; when two children share a key, only the first child will be used.`

benkeen commented:
All fixed except for the last React error you spotted. I'll fix that up now. The code shouldn't allow adding the same item multiple times as it does right now; I spoke to Sean and I'm going to add an flash effect on the input field to indicate there's a problem. This is very much a fringe case, so that form of error should be sufficient.

benkeen commented:
k, here we go. Had to do a hack to get the error highlight blink to work, but it was between two evils: using jQuery or plain CSS, so I vouched for the latter.

benkeen commented:
Commits squashed. The failing unit test was unrelated to this ticket. Let's see what Travis says this time.

garrensmith commented:
+1. This is great work @benkeen. Merge it.

benkeen commented:
Merged as 392dc52
